Add input checking and improve test coverage

jamesmbaazam opened the pull request:
This PR addresses comments raised in #79 and fixes #113.

sbfnk commented:
Great! Note the decrease in code coverage - can we add a test for this?

jamesmbaazam commented:
Good point. That should be straightforward. Thanks for taking a look.

pratikunterwegs reviewed (changes requested):
Thanks for making these changes @jamesmbaazam. I've found a few places where some more input checks would be useful. These are internal functions so I appreciate that the inputs are likely to be well known and controlled, but these extra checks would still help to isolate issues during development.

Regarding the tests, some of them have expectations on (some of) the value limits, and it would be good to have these tests for all functions that return values. For functions that take a vector and return outputs, it would also be good to test that the input and output lengths correspond as expected.

jamesmbaazam commented on R/borel.r:
```diff
@@ -1,12 +1,15 @@
 ##' @return probability mass.
 ##' @author Sebastian Funk
 dborel <- function(x, mu, log = FALSE) {
   if (x < 1) stop("'x' must be greater than 0")
```
I've now wrapped it in an `any()` call and updated the message here 2d5894385007e321c63a56819de79609c39db3a6.

jamesmbaazam commented on R/borel.r:
```diff
@@ -1,12 +1,15 @@
 ##'
 ##' @param x vector of integers.
-##' @param mu mu parameter.
+##' @param mu mu parameter (the Poisson mean).
```
`mu` is any non-negative number less than `Inf`.

jamesmbaazam commented on R/borel.r:
```diff
@@ -1,12 +1,15 @@
 ##' @author Sebastian Funk
 dborel <- function(x, mu, log = FALSE) {
   if (x < 1) stop("'x' must be greater than 0")
+  if (mu <= 0 || is.infinite(mu)) {
```
I've done that here 7ced0085e7e6773ab7a24cb39f67a2b762523403.

jamesmbaazam commented on R/borel.r:
```diff
@@ -16,11 +19,14 @@ dborel <- function(x, mu, log = FALSE) {
 ##'     if this number is reached
 ##' @return vector of random numbers
 ##' @author Sebastian Funk
 rborel <- function(n, mu, infinite = Inf) {
```
I've done that here 7ced0085e7e6773ab7a24cb39f67a2b762523403.

jamesmbaazam commented on R/borel.r:
```diff
@@ -16,11 +19,14 @@ dborel <- function(x, mu, log = FALSE) {
 ##' @return vector of random numbers
 ##' @author Sebastian Funk
 rborel <- function(n, mu, infinite = Inf) {
+  if (mu <= 0 || is.infinite(mu)) {
```
I've done that here 7ced0085e7e6773ab7a24cb39f67a2b762523403.

pratikunterwegs commented:
Thanks for looking at this review again @jamesmbaazam - I had forgotten about it entirely. Just a note that the commit hashes are not linking to the diff view, but that's no big deal. I'm just wondering why this PR is active now - my impression was that {bpmodels} was being superseded by {epichains} - will these changes be reflected there too?

jamesmbaazam commented on R/utils.r:
```diff
@@ -51,6 +51,7 @@ rgen_length <- function(n, x, prob) {
 #' @examples
 #' rnbinom_mean_disp(n = 5, mn = 4, disp = 2)
 rnbinom_mean_disp <- function(n, mn, disp) {
+  if (disp < 1) stop("'disp' must be at least 1")
```
I've added checks for all the inputs here 9eadc177e7f9c8526f53c23ed7b43ee02b0f94ab.

jamesmbaazam commented on R/utils.r:
```diff
@@ -51,6 +51,7 @@ rgen_length <- function(n, x, prob) {
 #' @examples
 #' rnbinom_mean_disp(n = 5, mn = 4, disp = 2)
 rnbinom_mean_disp <- function(n, mn, disp) {
```
Fixed here 9eadc177e7f9c8526f53c23ed7b43ee02b0f94ab.

jamesmbaazam commented:
> Just a note that the commit hashes are not linking to the diff view, but that's no big deal. 

They are local hashes. I'm yet to push them. They'll link up when I've pushed them.

> I'm just wondering why this PR is active now - my impression was that {bpmodels} was being superseded by {epichains} - will these changes be reflected there too?

Valid question. {epichains} shares these util functions and will benefit from this exercise. My intention is to port over these changes.

jamesmbaazam commented on tests/testthat/tests-sim.r:
```diff
@@ -18,6 +19,17 @@ test_that("Chains can be simulated", {
+  tf <- 3
+  set.seed(12)
+  expect_true(
+    all(chain_sim(
+      n = 2, offspring = "pois", "size", lambda = 0.9,
+      tree = TRUE, serial = function(n) {
+        rlnorm(n, meanlog = 0.58, sdlog = 1.58)
+      }, tf = tf
+    )$time < tf)
```
That's a valid point. I've made the change here 6c0fcbcec63c951c92852a07183390d7699a2c5f.

jamesmbaazam commented on tests/testthat/test-utils.R:
```diff
@@ -0,0 +1,20 @@
+  expect_length(rbinom_size(n = 1, x = c(1, 2, 3), prob = 0.3), 3)
+  expect_identical(complementary_logprob(x = 0), -Inf)
+  expect_identical(complementary_logprob(x = -Inf), 0)
+  expect_lt(complementary_logprob(x = -0.1), 0)
```
Not sure I understand. The test on line 7 above tests for the upper value when `x = -0.1` as you say.

jamesmbaazam commented:
Thanks for reviewing this @pratikunterwegs. I've addressed your comments and will merge this as it's been inactive for a pretty long time because it was deprioritized. I'll port over the changes to {epichains} where I will accept further reviews.
